File: mtgfish/sim/board.py
"""The board, as a person would look at it, sampled through a replay.

The log says what happened. It does not say what the table *looked like*, and
after fifty turns a log is a very long way of saying "your deck made five
thousand Ape tokens". Anyone who plays Magic reads a board in a second and a
log in ten minutes, so a replay that can only be read is a replay that mostly
is not.

So a replay is also a film: the whole visible position, sampled at every moment
that could have changed it, each sample tagged with the log entry it is current
as of. The viewer scrubs one and the other follows.

Three things make this affordable, because the naive version is 12 MB of JSON
for one game:

**Identity is stored once.** A permanent in a snapshot is an integer into
``BoardFilm.cards``; the name, art and parse verdict live there. Repeating them
for fifty permanents across six hundred snapshots is most of that 12 MB.

**Identical boards collapse.** Sampling on every interesting event gives about
3,700 samples of which roughly 600 differ - the rest are a step beginning and
ending with nothing between. A sample equal to the one before it extends that
one's reach instead of being stored, which loses nothing: the board really was
unchanged.

**Nothing is computed that the engine would not compute anyway.** Power and
toughness come from ``characteristics``, which is memoised per epoch, so a
watched game plays the same game at close to the same speed.
"""

from __future__ import annotations

from dataclasses import dataclass, field, replace
from typing import TYPE_CHECKING

from ..rules.enums import Zone
from ..rules.events import EventKind
from ..rules.gameobject import ObjectKind

if TYPE_CHECKING:
    from ..rules.game import Game

#: Events after which the board may look different. Generous on purpose: an
#: event that changed nothing costs one comparison and is then dropped, while
#: an event left out is a change the viewer never shows.
SAMPLE_AFTER: frozenset[EventKind] = frozenset(
    {
        EventKind.ZONE_CHANGE,
        EventKind.ENTERS_BATTLEFIELD,
        EventKind.LEAVES_BATTLEFIELD,
        EventKind.DIES,
        EventKind.TOKEN_CREATED,
        EventKind.CEASED_TO_EXIST,
        EventKind.CAST_SPELL,
        EventKind.SPELL_RESOLVED,
        EventKind.ABILITY_RESOLVED,
        EventKind.PUT_ON_STACK,
        EventKind.COUNTERED,
        EventKind.FIZZLED,
        EventKind.TAPPED,
        EventKind.UNTAPPED,
        EventKind.COUNTER_ADDED,
        EventKind.COUNTER_REMOVED,
        EventKind.ATTACHED,
        EventKind.UNATTACHED,
        EventKind.TRANSFORMED,
        EventKind.PHASED_OUT,
        EventKind.PHASED_IN,
        EventKind.CONTROL_CHANGED,
        EventKind.DAMAGE_DEALT,
        EventKind.COMBAT_DAMAGE_DEALT,
        EventKind.LIFE_GAINED,
        EventKind.LIFE_LOST,
        EventKind.LIFE_SET,
        EventKind.POISON_ADDED,
        EventKind.DREW_CARD,
        EventKind.DISCARDED,
        EventKind.MILLED,
        EventKind.ATTACKERS_DECLARED,
        EventKind.BLOCKERS_DECLARED,
        EventKind.REMOVED_FROM_COMBAT,
        EventKind.TURN_BEGAN,
        EventKind.STEP_BEGAN,
        EventKind.STEP_ENDED,
        EventKind.PLAYER_LOST,
    }
)


@dataclass(frozen=True, slots=True)
class BoardCardRef:
    """A distinct printed card - or token face - the viewer has to draw.

    ``oracle_id`` is empty for a token, which has no printing and therefore no
    art to fetch; the viewer draws those from the name and the type line.
    """

    name: str
    oracle_id: str = ""
    type_line: str = ""
    mana_cost: str = ""
    token: bool = False


@dataclass(slots=True)
class BoardCard:
    """One permanent, with everything needed to draw it and nothing else."""

    id: int
    card: int
    tapped: bool = False
    power: int | None = None
    toughness: int | None = None
    damage: int = 0
    counters: dict = field(default_factory=dict)
    sick: bool = False
    #: The player this creature is attacking, or -1. A separate field from
    #: ``attacking_permanent`` because player ids and object ids are both small
    #: integers - see ``rules.combat.Combat``.
    attacking: int = -1
    attacking_permanent: int = -1
    blocking: tuple[int, ...] = ()
    token: bool = False
    commander: bool = False
    face_down: bool = False
    phased_out: bool = False
    attached_to: int = -1
    #: True when no ability of this card was read. The whole reason to draw a
    #: board rather than describe it: a permanent that does nothing is visible
    #: as a blank the moment it lands, not inferred from a log that never
    #: mentions it again.
    unreadable: bool = False


@dataclass(slots=True)
class BoardSeat:
    """One player, and what they have."""

    player: int
    life: int
    hand: int
    library: int
    graveyard: int
    exile: int
    poison: int = 0
    out: bool = False
    permanents: list[BoardCard] = field(default_factory=list)
    #: The cards in hand and in the graveyard, when the viewer is allowed to
    #: see them. Empty in a replay, always: a replay that listed a hand would
    #: be showing information nobody at the table had. The sandbox is not a
    #: game - it is a bench with one operator - and shows everything.
    hand_cards: list[BoardCard] = field(default_factory=list)
    graveyard_cards: list[BoardCard] = field(default_factory=list)


@dataclass(slots=True)
class BoardStackItem:
    card: int
    controller: int
    #: Spell, or the text of the ability waiting to resolve.
    text: str = ""


@dataclass(slots=True)
class BoardSnapshot:
    """The whole visible position at one moment."""

    #: The log entry this board is current as of. The viewer scrubs the log and
    #: shows the last snapshot at or before wherever it is.
    frame: int
    turn: int
    phase: str
    step: str
    active: int
    seats: list[BoardSeat] = field(default_factory=list)
    #: Top of the stack last, the way it resolves.
    stack: list[BoardStackItem] = field(default_factory=list)

    def same_board_as(self, other: BoardSnapshot) -> bool:
        """Whether nothing a viewer could see has changed.

        ``frame`` is excluded on purpose - that is the whole point - and so are
        turn, phase and step, which are shown from the log's own heading.
        """
        return self.seats == other.seats and self.stack == other.stack


@dataclass(slots=True)
class BoardFilm:
    """Every distinct card in the game, and the board over time."""

    cards: list[BoardCardRef] = field(default_factory=list)
    snapshots: list[BoardSnapshot] = field(default_factory=list)

    def at_frame(self, frame: int) -> BoardSnapshot | None:
        """The board as it stood at a log entry - the last one at or before it."""
        found = None
        for snapshot in self.snapshots:
            if snapshot.frame > frame:
                break
            found = snapshot
        return found


class BoardRecorder:
    """Samples the board, attached to a game as a second observer.

    Deliberately read-only: it asks the engine only for things the engine has
    already computed, and never asks it to do anything. A watched game has to
    be the same game as an unwatched one, or the film shows a board the
    statistics did not come from.
    """

    __slots__ = ("film", "open_zones", "_index", "_previous")

    def __init__(self, *, open_zones: bool = False) -> None:
        self.film = BoardFilm()
        #: Whether to record the contents of hands and graveyards rather than
        #: only their size. False for a replay, and the default is False so
        #: that leaking a hand has to be asked for.
        self.open_zones = open_zones
        #: card reference -> its index in ``film.cards``.
        self._index: dict[BoardCardRef, int] = {}
        self._previous: BoardSnapshot | None = None

    def __call__(self, game: Game, event) -> None:
        if event.kind not in SAMPLE_AFTER:
            return
        self.sample(game)

    def sample(self, game: Game) -> None:
        """Take one snapshot, unless it would repeat the last one."""
        snapshot = self._snapshot(game)
        previous = self._previous
        if previous is not None and snapshot.same_board_as(previous):
            # The board really is unchanged, so the earlier snapshot simply
            # stays current for longer. Storing it again would be storing the
            # same picture twice.
            return
        self.film.snapshots.append(snapshot)
        self._previous = snapshot

    # -- building a snapshot ------------------------------------------------

    def _snapshot(self, game: Game) -> BoardSnapshot:
        combat = game.combat
        seats = [self._seat(game, player, combat) for player in game.players]
        return BoardSnapshot(
            frame=len(game.log.entries),
            turn=game.turn,
            phase=game.phase.name,
            step=game.step.name,
            active=int(game.active_player),
            seats=seats,
            stack=[
                item
                for object_id in game.stack
                if (item := self._stack_item(game, object_id)) is not None
            ],
        )

    def _hidden(self, game: Game, object_ids) -> list[BoardCard]:
        """Cards in a zone a viewer is allowed to look into."""
        if not self.open_zones:
            return []
        out = []
        for object_id in object_ids:
            obj = game.objects.get(object_id)
            if obj is not None:
                out.append(self._permanent(game, obj, None))
        return out

    def _seat(self, game: Game, player, combat) -> BoardSeat:
        return BoardSeat(
            player=int(player.id),
            life=player.life,
            hand=len(player.hand),
            library=len(player.library),
            graveyard=len(player.graveyard),
            exile=sum(
                1
                for object_id in game.exile
                if (obj := game.objects.get(object_id)) is not None
                and obj.owner == player.id
            ),
            poison=player.poison,
            out=player.has_lost,
            permanents=[
                self._permanent(game, obj, combat)
                for obj in game.permanents(player.id)
            ],
            hand_cards=self._hidden(game, player.hand),
            graveyard_cards=self._hidden(game, player.graveyard),
        )

    def _permanent(self, game: Game, obj, combat) -> BoardCard:
        chars = game.characteristics(obj)
        attacking = -1
        attacking_permanent = -1
        blocking: tuple[int, ...] = ()
        if combat is not None:
            if obj.id in combat.attacking:
                attacking = int(combat.attacking[obj.id])
                attacking_permanent = int(
                    combat.attacking_permanent.get(obj.id, -1)
                )
            blocked = combat.blocking.get(obj.id)
            if blocked:
                blocking = tuple(int(i) for i in blocked)
        return BoardCard(
            id=int(obj.id),
            card=self._card_index(obj, chars),
            tapped=obj.tapped,
            power=chars.power,
            toughness=chars.toughness,
            damage=obj.damage,
            counters={k: v for k, v in obj.counters.items() if v},
            sick=obj.summoning_sick and chars.is_creature,
            attacking=attacking,
            attacking_permanent=attacking_permanent,
            blocking=blocking,
            token=obj.kind is ObjectKind.TOKEN,
            commander=obj.is_commander,
            face_down=obj.face_down,
            phased_out=obj.phased_out,
            attached_to=int(obj.attached_to) if obj.attached_to else -1,
            unreadable=bool(chars.abilities)
            and all(a.unparsed for a in chars.abilities),
        )

    def _stack_item(self, game: Game, object_id) -> BoardStackItem | None:
        obj = game.objects.get(object_id)
        if obj is None:
            return None
        chars = game.characteristics(obj)
        return BoardStackItem(
            card=self._card_index(obj, chars),
            controller=int(obj.controller),
            text=(obj.ability.text if obj.ability is not None else "") or "",
        )

    def _card_index(self, obj, chars) -> int:
        """Intern this card, so a snapshot stores an integer and not a name."""
        card = getattr(obj, "card", None)
        ref = BoardCardRef(
            name=chars.name or getattr(card, "name", "") or "",
            oracle_id=getattr(card, "oracle_id", "") or "",
            type_line=str(chars.type_line),
            mana_cost=str(chars.mana_cost) if chars.mana_cost else "",
            token=obj.kind is ObjectKind.TOKEN,
        )
        index = self._index.get(ref)
        if index is None:
            index = len(self.film.cards)
            self._index[ref] = index
            self.film.cards.append(ref)
        return index

File: mtgfish/ui/web/board.js
"use strict";

/* The replay as a board, rather than as a document.
 *
 * A log says what happened; it does not say what the table looked like. After
 * fifty turns that difference is the whole thing - anyone who plays Magic
 * reads a board in a second and a log in ten minutes, and "your deck made five
 * thousand Ape tokens" is a sentence nobody should have to read to learn that
 * their deck made five thousand Ape tokens.
 *
 * Three things this view is for, in order:
 *
 *   what is on the battlefield, whose it is, and what is tapped;
 *   what is attacking, and whom;
 *   which permanents the parser could not read - visible as a blank the moment
 *   it lands, rather than inferred from a log that never mentions it again.
 *
 * The boards live on the Python side. The page is told where each one sits in
 * the log and fetches them one at a time, because the whole film is tens of
 * megabytes for one game and only ever one of them is on screen. */

const Board = (() => {
  let cards = [];        // the card dictionary: art, type line, parse status
  let frames = [];       // log frame index of each board, ascending
  let current = -1;      // which board is loaded
  let onJump = null;     // told when the viewer moves, so the log can follow

  /* Cards are stacked by name. A board with 5,000 Ape tokens drawn one card
   * per token is not a board, it is a denial of service - and the thing worth
   * seeing is the number anyway. */
  const STACK_FROM = 2;

  function reset(payload, jumpHandler) {
    cards = payload.board_cards || [];
    frames = payload.board_frames || [];
    current = -1;
    onJump = jumpHandler;
    const host = document.getElementById("replayboardview");
    host.innerHTML = "";
    const slider = document.getElementById("boardscrub");
    slider.max = Math.max(0, frames.length - 1);
    slider.value = 0;
    slider.disabled = frames.length === 0;
    document.getElementById("boardpanel").hidden = frames.length === 0;
    if (frames.length) show(0);
  }

  /* Which board is current at a log position: the last one at or before it.
   * Binary search because a long game has a couple of thousand of them and
   * this runs on every scroll of the log. */
  function indexForFrame(frame) {
    let low = 0;
    let high = frames.length - 1;
    let found = 0;
    while (low <= high) {
      const mid = (low + high) >> 1;
      if (frames[mid] <= frame) { found = mid; low = mid + 1; } else { high = mid - 1; }
    }
    return found;
  }

  async function showAtFrame(frame) {
    const index = indexForFrame(frame);
    if (index !== current) await show(index);
  }

  async function show(index) {
    if (!frames.length) return;
    index = Math.max(0, Math.min(index, frames.length - 1));
    const board = await call("replay_board", index);
    if (board.error) return;
    current = index;
    document.getElementById("boardscrub").value = index;
    draw(board);
  }

  function step(delta) {
    show(current + delta);
    if (onJump) onJump(frames[Math.max(0, Math.min(current + delta, frames.length - 1))]);
  }

  /* ------------------------------------------------------------- drawing */

  function draw(board) {
    document.getElementById("boardwhere").textContent =
      `turn ${board.turn} · ${board.phase.toLowerCase().replace(/_/g, " ")}`
      + ` / ${board.step.toLowerCase().replace(/_/g, " ")}`;
    drawInto(document.getElementById("replayboardview"), cards, board);
  }

  /* The whole board, into whatever element asked for it.
   *
   * The sandbox draws with this too. A bench that looked different from a
   * replay would be a second thing to learn for no reason, and the two do
   * differ in exactly one way: on a bench there is one operator who put every
   * card there themselves, so hands and graveyards are shown rather than
   * counted. Everything else is the same board. */
  function drawInto(host, table, board, options) {
    options = options || {};
    const was = cards;
    cards = table || [];
    try {
      host.innerHTML = "";
      const names = options.names || {};
      board.seats.forEach((seat) => {
        if (names[seat.player] === undefined) names[seat.player] = `P${seat.player}`;
      });

      const grid = el("div", "boardtable");
      board.seats.forEach((seat) => grid.appendChild(drawSeat(seat, board, names, options)));
      host.appendChild(grid);

      if (board.stack.length) host.appendChild(drawStack(board));
    } finally {
      // Only the replay keeps a card table between calls; a one-off draw must
      // not leave the scrubber pointing at somebody else's cards.
      if (options.keep !== true) cards = was;
    }
  }

  function drawSeat(seat, board, names, options) {
    const node = el("div", `boardseat seat-${seat.player}`
      + (seat.player === board.active ? " active" : "")
      + (seat.out ? " out" : ""));

    const head = el("div", "seathead");
    head.appendChild(el("span", "seatwho", names[seat.player]));
    head.appendChild(el("span", "seatlife", `${seat.life}`));
    const counts = el("span", "seatcounts");
    counts.appendChild(el("span", "count", `${seat.hand} hand`));
    counts.appendChild(el("span", "count", `${seat.library} library`));
    counts.appendChild(el("span", "count", `${seat.graveyard} yard`));
    if (seat.exile) counts.appendChild(el("span", "count", `${seat.exile} exile`));
    if (seat.poison) counts.appendChild(el("span", "count bad", `${seat.poison} poison`));
    head.appendChild(counts);
    if (seat.player === board.active) head.appendChild(el("span", "seatturn", "their turn"));
    if (seat.out) head.appendChild(el("span", "seatturn bad", "out"));
    node.appendChild(head);

    if (!seat.permanents.length) {
      node.appendChild(el("div", "boardempty", "nothing on the battlefield"));
      openZones(node, seat, board, names);
      return node;
    }

    /* Lands read as a block and everything else as individuals, which is how
     * a player looks at a board: you count the mana, then you read the cards. */
    const lands = [];
    const rest = [];
    seat.permanents.forEach((p) => {
      (isLand(p) ? lands : rest).push(p);
    });

    if (rest.length) node.appendChild(drawRow(rest, board, names));
    if (lands.length) node.appendChild(drawRow(lands, board, names, "lands"));
    openZones(node, seat, board, names);
    return node;
  }

  /* Hand and graveyard, when the caller is allowed to see them. A replay never
   * sends these - it would be showing information nobody at the table had. */
  function openZones(node, seat, board, names) {
    [["hand", seat.hand_cards], ["graveyard", seat.graveyard_cards]].forEach(
      ([label, contents]) => {
        if (!contents || !contents.length) return;
        node.appendChild(el("div", "zonename", label));
        const row = drawRow(contents, board, names, "openzone");
        node.appendChild(row);
      }
    );
  }

  function isLand(permanent) {
    const card = cards[permanent[1]];
    return card && /\bLand\b/.test(card.type_line);
  }

  /* One row of permanents, identical ones stacked into a single card with a
   * count on it. */
  function drawRow(permanents, board, names, extra) {
    const row = el("div", "boardrow" + (extra ? " " + extra : ""));
    const groups = new Map();
    permanents.forEach((p) => {
      // Anything with its own state - tapped, damaged, countered, attacking -
      // is worth seeing on its own; only plain duplicates collapse.
      const key = plain(p) ? `${p[1]}|${p[2]}` : `solo-${p[0]}`;
      const group = groups.get(key);
      if (group) group.push(p); else groups.set(key, [p]);
    });
    groups.forEach((group) => row.appendChild(drawCard(group, board, names)));
    return row;
  }

  function plain(permanent) {
    return permanent.length < 7;   // no counters, not attacking, not blocking
  }

  /* Art comes from Scryfall, over the network, and so is the one part of this
   * view that can simply not arrive: a token has no printing to fetch, the
   * desktop app may be offline, and Scryfall may be having a day. A card whose
   * art is missing still has to read as that card, so the placeholder keeps
   * the name and the frame and only loses the picture - a broken-image icon in
   * a row of eighty is worse than no picture at all. */
  function artFor(card) {
    if (!card.art) return el("div", "bart token");
    const img = el("img", "bart");
    img.loading = "lazy";
    img.src = card.art;
    img.alt = "";
    img.addEventListener("error", () => {
      const blank = el("div", "bart token");
      if (img.parentNode) img.parentNode.replaceChild(blank, img);
    });
    return img;
  }

  function drawCard(group, board, names) {
    const permanent = group[0];
    const card = cards[permanent[1]] || { name: "?", type_line: "", status: "ok" };
    const flags = permanent[2];
    const extra = permanent[6] || {};

    const node = el("div", "bcard"
      + ((flags & 1) ? " tapped" : "")
      + ((flags & 2) ? " sick" : "")
      + ((flags & 32) ? " phased" : "")
      + (card.status === "blank" || (flags & 64) ? " unread" : "")
      + (card.status === "partial" ? " partial" : "")
      + (extra.a !== undefined ? " attacking" : "")
      + (extra.b ? " blocking" : ""));

    node.appendChild(artFor(card));
    node.appendChild(el("div", "bname", card.name));

    const marks = el("div", "bmarks");
    if (group.length > STACK_FROM - 1 && group.length > 1) {
      marks.appendChild(el("span", "bcount", `x${group.length}`));
    }
    if (permanent[3] !== null && permanent[3] !== undefined) {
      const pt = el("span", "bpt", `${permanent[3]}/${permanent[4]}`);
      if (permanent[5]) pt.classList.add("hurt");
      marks.appendChild(pt);
    }
    if (permanent[5]) marks.appendChild(el("span", "bdmg", `${permanent[5]} dmg`));
    Object.entries(extra.c || {}).forEach(([kind, n]) => {
      marks.appendChild(el("span", "bcounter", `${n}×${kind}`));
    });
    if (flags & 8) marks.appendChild(el("span", "bcmd", "CMD"));
    if (marks.childNodes.length) node.appendChild(marks);

    if (extra.a !== undefined) {
      const at = extra.ap !== undefined && extra.ap >= 0
        ? "a permanent" : (names[extra.a] || `P${extra.a}`);
      node.appendChild(el("div", "battack", `attacking ${at}`));
    }
    if (extra.b) node.appendChild(el("div", "battack block", "blocking"));

    node.title = describe(card, permanent, group.length);
    node.addEventListener("click", () => zoom(card));
    return node;
  }

  function describe(card, permanent, count) {
    const lines = [card.name + (count > 1 ? ` (x${count})` : ""), card.type_line];
    if (permanent[2] & 1) lines.push("tapped");
    if (permanent[2] & 2) lines.push("summoning sick");
    if (card.status === "blank") {
      lines.push("THE PARSER READ NOTHING OF THIS CARD - it does nothing at all");
    } else if (card.status === "partial") {
      lines.push("partly read: " + (card.unread || []).join(" / "));
    }
    return lines.join("\n");
  }

  function drawStack(board) {
    const node = el("div", "boardstack");
    node.appendChild(el("div", "zonename", "Stack - top resolves first"));
    const row = el("div", "boardrow");
    board.stack.slice().reverse().forEach((item) => {
      const card = cards[item.card] || { name: "?" };
      const entry = el("div", `bcard onstack seat-${item.controller}`);
      entry.appendChild(artFor(card));
      entry.appendChild(el("div", "bname", card.name));
      if (item.text) entry.title = item.text;
      entry.addEventListener("click", () => zoom(card));
      row.appendChild(entry);
    });
    node.appendChild(row);
    return node;
  }

  /* The full card, because the board is drawn from art crops and sometimes the
   * question is "what does this actually say". */
  function zoom(card) {
    const host = document.getElementById("boardzoom");
    host.innerHTML = "";
    host.hidden = false;
    if (card.image) {
      const img = el("img");
      img.src = card.image;
      img.alt = card.name;
      img.addEventListener("error", () => img.remove());
      host.appendChild(img);
    }
    const info = el("div", "zoominfo");
    info.appendChild(el("h3", null, card.name));
    info.appendChild(el("div", "dim", card.type_line));
    if (card.status === "blank") {
      info.appendChild(el("div", "reading fail",
        "The parser read nothing of this card. On the battlefield it is a blank: "
        + "every number this run produced understates it."));
    } else if (card.status === "partial") {
      info.appendChild(el("div", "reading fail",
        "Partly read. These abilities never fired: " + (card.unread || []).join("; ")));
    } else if (card.token) {
      info.appendChild(el("div", "dim", "A token - no printing, so no art."));
    } else {
      info.appendChild(el("div", "reading ok", "The parser read this card completely."));
    }
    const close = el("button", "ghost small", "Close");
    close.addEventListener("click", () => { host.hidden = true; });
    info.appendChild(close);
    host.appendChild(info);
  }

  return { reset, show, showAtFrame, step, indexForFrame, drawInto,
           get count() { return frames.length; },
           get frames() { return frames; },
           get current() { return current; } };
})();
